client.js

```diff
@@ -100,75 +112,94 @@ RingpopClient.prototype.destroy = function destroy(callback) {
     if (this.isChannelOwner) {
         this.tchannel.close(callback);
     }
+
+    this.timers.clearTimeout(this.wedgedTimer);
+    this.isDestroyed = true;
 };
 
 /* jshint maxparams: 5 */
 RingpopClient.prototype._request = function _request(opts, endpoint, head, body, callback) {
     var self = this;
 
-    if (this.subChannel && this.subChannel.destroyed) {
-        process.nextTick(function onTick() {
-            callback(ChannelDestroyedError({
-                endpoint: endpoint,
-                channelType: 'subChannel'
-            }));
-        });
-        return;
+    if (!this.wedgedTimer) {
+        this._scheduleWedgedTimer();
     }
 
-    if (this.subChannel &&
-            this.subChannel.topChannel &&
-            this.subChannel.topChannel.destroyed) {
+    var request = new ClientRequest(this, opts, endpoint, head, body, onSend);
+
+    // Evaluate if Ringpop should start applying backpressure on itself.
+    // Allow only a certain amount of potentially wedged requests to exist.
+    // It is highly unlikely that this'll ever be triggered, but it's best
+    // not to issue an unbounded number of wedged requests.
+    var inflightCurrent = Object.keys(this.requestsById).length;
+    var inflightMax = this.config.get('inflightClientRequestsLimit');
+    if (inflightCurrent > inflightMax) {
         process.nextTick(function onTick() {
-            callback(ChannelDestroyedError({
-                endpoint: endpoint,
-                channelType: 'topChannel'
+            callback(ClientRequestLimitError({
+                inflightCurrent: inflightCurrent,
+                inflightMax: inflightMax,
+                endpoint: endpoint
             }));
         });
         return;
     }
 
-    if (!opts || !validateHostPort(opts.host)) {
-        callback(InvalidHostPortError({
-            hostPort: String(opts && opts.host)
-        }));
-        return;
+    // Registering the request by its ID must happen at the precise
+    // moment before we know it will be "inflight."
+    this.requestsById[request.id] = request;
+    request.send();
+
+    function onSend() {
+        delete self.requestsById[request.id];
+        var args = Array.prototype.splice.call(arguments, 0);
+        callback.apply(null, args);
     }
+};
 
-    self.subChannel.waitForIdentified({
-        host: opts.host
-    }, function onIdentified(err) {
-        if (err) {
-            callback(err);
-            return;
+RingpopClient.prototype._scanForWedgedRequests = function _scanForWedgedRequests() {
+    var wedgedRequests = [];
+
+    // Find all requests that have not been timed-out by TChannel.
+    // Induce Ringpop timeout for wedged requests and cancel them.
+    var requestIds = Object.keys(this.requestsById);
+    for (var i = 0; i < requestIds.length; i++) {
+        var request = this.requestsById[requestIds[i]];
+        var timeInflight = Date.now() - request.timestamp;
```

Comment:
nit: should Date.now move outside of loop to `var now = Date.now();` and be used as such?